### ISSUE TITLE
Headless Client Connection

### ISSUE THREAD

**Issue by ToasterBR** [Dahlgren/node-arma-server#21]
Hi @Dahlgren!

Recently my HC stopped to connect into my main server using Web Admin (which uses this module).

First, I suspected the problem could be the params order in the `src/headless.js`, so I changed to match some of my old `.bat` files. It didn´t work.

Then, I suspected it could be the params in the `server.cfg`, but the file generated by the module was correct.

I tried to start a server without password but I had no joy either. 

After everything, I built a mission without any mods. Added some HCs and set them to playable with names HC1 to HC3. Uploaded my mission to the server and set the server to use a HC. Still no connection.

The funny thing is that the HC only shows `23:14:30 Dedicated client created.`
No connection error, no password error, nothing.

If I make a `.bat` and set it to connect using:
`arma3server.exe -client -connect=127.0.0.1 -port=2312 -password=password`

It works. ¯\_(ツ)_/¯

I uploaded the testing mission in this link: [headless_testing.VR.pbo.zip](https://github.com/Dahlgren/node-arma-server/files/1030632/headless_testing.VR.pbo.zip)

Thanks in advance! This is driving me crazy haha

PS: I'm using the `feature/steam-workshop-mods` branch.

**Comment by Dahlgren** [comment]
Thanks, I'll look into it! Hopefully today, otherwise during the weekend. We're using HC a lot so this probably impacts us as well. I think we observed something similar a week ago or so.

Is this with the 1.70 hotfix or just 1.70? And I'm guessing you're using Windows based on `.bat` and `arma3server.exe` :)

Did you use mods from Steam Workshop while testing?

Thanks for the test mission!

**Comment by ToasterBR** [comment]
Hey!
Yes, I applied the hotfix and I'm using in a Win10 machine.

I tried with and without Steam Workshop mods.

Ah! I tried with x86 and x64 versions of the Dedicated Server.

Thanks!

**Comment by Dahlgren** [comment]
I seem to be unable to replicate your issue. The headless client launched from the manager is able to join the server with the proper printouts. Anything funky in the server rpt? Any stray pbo files in the Arma 3 addons folder? If the headless client is unable to join the server it won't show anything more than `Dedicated client created.`

**Server**
````
11:55:54 Dedicated host created.
11:56:01 Host identity created.
11:56:01 Game Port: 2302, Steam Query Port: 2303
11:56:02 Player headlessclient connecting.
11:56:02 Mission headless_testing.VR read from bank.
11:56:03 Player headlessclient connected (id=HC5008).
````

**Headless**
````
11:55:53 Dedicated client created.
11:56:02 Client connected: 127.0.0.1:2302
11:56:03  > Player headlessclient connecting
11:56:04 Loading profile "HC1"
````

Could you check with [ProcessExplorer](https://technet.microsoft.com/en-us/sysinternals/processexplorer.aspx) if all parameters are passed as they should?

It should have at least `-connect=127.0.0.1`, `-client`, `-port=${server port}` and `-password=${server password if any}`

**Comment by ToasterBR** [comment]
Found the problem.

I don't know if it's a bug or if it should be reported here, but the problem was passing -filePatching as parameter in config.js. I needed that in a mod I was using to give auto zeus for admins. ([@slz](http://www.armaholic.com/page.php?id=30621))

![Bad parameter?](http://i.imgur.com/I7wnadG.png)

With that on, the module tries to pass the same parameter to the HC:

![Bad HC](http://i.imgur.com/fNxvRuN.png)

Without `-filePatching`, the HC connects successfully.

Thanks to let me know about processExplorer. Was very helpful!

**Comment by Dahlgren** [comment]
Might be because of server disallowing file patching by default. Should probably create two separate config entries, one for extra parameters to server and one for headless client. Maybe also add an option to configure the server to allow clients (and/or headless) to enable file patching.

````
allowedFilePatching = 0;
Prevent or allow file patching for the clients (including the HC) (since Arma 3 1.49+)
0 is no clients (default)
1 is Headless Clients only
2 is all clients
````

If you want to load extra config files without filePatching you can just pack it into a pbo and load it with a mod folder

**Comment by ToasterBR** [comment]
You're right.

I added the `allowedFilePatching = 1;` in the `templates/config.tpl` and it worked! :)

Thanks!

**Comment by Dahlgren** [comment]
I've started refactoring the handling of config file generation and will add the `allowedFilePatching` as a possible value once that is done. It still needs to be added to the web admin once that's done.

**Comment by Dahlgren** [comment]
Should be fixed by Dahlgren/arma-server-web-admin/pull/81

**Comment by ToasterBR** [comment]
Can confirm.  It was fixed! Thanks @Dahlgren!
